### DISCUSSION TITLE
anonymous `using` disposable object

### DISCUSSION THREAD

**Discussion by skyoxZ** [dotnet/csharplang#6005] [Language Ideas]
Proposal:

```csharp
string path = (using Process.GetCurrentProcess()).MainModule.FileName;
```

generates code similar to:

```csharp
using var xxxx1234 = Process.GetCurrentProcess();
string path = xxxx1234.MainModule.FileName;
```

and also consider `(using new SomeDisposable()).Xxx` and `(await using XXX).Xxx`

**Comment by Joe4evr** [comment]
#2593

**Comment by skyoxZ** [comment]
@Joe4evr Not a duplicate, I think. My `using` expression results an object.

**Comment by svick** [comment]
My problem with this is that it's not clear when exactly the object is going to be disposed. And with disposal, clarity is important.

**Reply by TahirAhmadov** [reply]
Isn't it easy to just define the rule at "end of current statement"? (meaning `;`)

**Reply by svick** [reply]
@TahirAhmadov 

I think that would not be clear. Consider small refactoring of the code in the original post:

```c#
var module = (using Process.GetCurrentProcess()).MainModule;
string path = module.FileName;
```

Is this code now problematic? (It seems disposing the `Process` does not actually invalidate the `ProcessModule` here, but I'm sure you can imagine similar cases where that would happen.)

Or consider the following code:

```c#
using (var jsonReader = new JsonTextReader(using new StringReader(json)))
{
    return new JsonSerializer().Deserialize(jsonReader);
}
```

This code would work, since the "current statement" for the `using` expression is the `using` statement, which ends at the `}`, correct? And changing the code to use C# 8 `using` declaration would break it? That doesn't seem desirable to me.
